README.md
Clasificación de vocales con redes neuronales

Descripción del problema
Este trabajo tiene como objetivo el reconocimiento de caracteres escritos a mano, en particular de las vocales en minúscula, utilizando algoritmos simples de clasificación no lineal con redes neuronales. 
Metodología
Base de datos
La base de datos utilizada para el proyecto fue la  EMNIST: an extension of MNIST to handwritten letters” (https://arxiv.org/abs/1702.05373v1) .


La famosa base de datos MNIST se derivó de un conjunto de datos más grande conocido como NIST Special Database 19 que contiene imágenes con dimensiones de 128 x 128 pixeles de dígitos, letras escritas a mano en mayúsculas y minúsculas. Una variante del conjunto de datos NIST completo, denominado MNIST extendido (EMNIST),  sigue el mismo paradigma de conversión utilizado para crear el conjunto de datos MNIST. El resultado es un conjunto de conjuntos de datos que constituyen tareas de clasificación más difíciles que involucran letras y dígitos.

Para fines de este trabajo, se procedió a clasificar de manera manual 500 imágenes de cada vocal en minúscula, siendo un total de 2500 imágenes. 

De manera más específica, se separaron las imágenes por carpetas para cada vocal, y adicionalmente se dividió la cantidad de imágenes para entrenamiento (60%), validación (20%) y pruebas (20%). 

Por lo consiguiente, se obtuvieron 1500 imágenes para entrenamiento (300 de cada vocal), 500 para validación (100 de cada vocal) y 500 para pruebas (100 de cada vocal).  Las imágenes para entrenamiento fueron guardadas en una sola carpeta que incluyera a todas las vocales, sin embargo, se utilizó una nomenclatura de tal manera que las 300 primeras imágenes fueran correspondientes a la letra a, las siguientes 300 imágenes fueran correspondientes a la letra e, y así sucesivamente hasta llegar a la vocal u, esto con el objetivo de poder realizar posteriormente la etiquetación manual para realizar el aprendizaje supervisado. Un procedimiento similar fue realizado para las imágenes de validación y pruebas, con la diferencia de que la cantidad de imágenes por cada vocal fue de 100. 

Método
De primera instancia, para el procesamiento de las imágenes, éstas fueron reducidas a un 25% de su tamaño, es decir a 32 x 32 pixeles y convertidas a una escala de grises para así obtener una matriz de sus valores del 0 al 255, siendo 0 el color negro y el 255 el color blanco. Posteriormente, cada valor fue dividido entre 255.

Una vez realizada el procesamiento de imágenes, se procedió a realizar el etiquetado con una función en Python. Las etiquetas por cada letra corresponderían a un rango de 0 a 4, donde el número 0 representaría a la vocal a, el número 1 correspondería a la vocal e y la misma analogía para las vocales consecuentes. 

 Cada matriz de imagen fue guardada como un elemento de una lista, de acuerdo a como habían sido guardadas en su directorio correspondiente. De esta manera, los primeros 300 elementos correspondían a matrices de imágenes de la vocal a, por lo que se creó una lista alterna de etiquetas en la que los primeros 300 elementos fueran el número ‘0’. Siguiendo la analogía las siguientes etiquetas serían 300 elementos ‘1’ para la letra e, 300 elementos ‘2’ para la letra i, 300 elementos ‘3 para la letra o y finalmente 300 elementos ‘4’ para la letra u. 

De manera similar se creó una lista cuyo contenido sería la clase de cada imagen, definiendo las clases como ‘a’, ‘e’, ‘i’, ‘o’, y ‘u’.


Finalmente, una vez obtenidas las tres listas con las imágenes, etiquetas y clases, se procedió a unirlas en una sola lista, para de esta manera poder ordenar aleatoriamente los elementos y realizar la fase de entrenamiento. 


Cabe mencionar que el mismo procedimiento fue realizado para el conjunto de imágenes para validación y para las pruebas, cada una de ellas con su correspondiente etiqueta y clase. Adicionalmente, los conjuntos ordenado aleatoriamente fueron guardados en archivos para posteriormente cargarlos y procesarlos nuevamente. 

Arquitectura de la red
Para la construcción del modelo, se utilizó la biblioteca de Tensorflow Keras, una biblioteca de código abierto especializada para desarrollar modelos de aprendizaje automático. 

Esta biblioteca contiene un módulo para la implementación de modelos de redes neuronales. En este trabajo se utilizó el modelo secuencial proporcionado por la biblioteca, que es una pila lineal  de capas. 

Para la resolución de este problema, se utilizaron tres capas, una con 1024 neuronas de entrada correspondientes al número de pixeles por imagen,  otra capa con 128 neuronas con funciones de activación sigmoidales y por último, una capa de salida con 5 neuronas correspondientes a las 5 clases de vocales y con funciones de activación softmax. 

Optimización 
Para la funciones de optimización, se eligió la función “Adam”, debido a que fue la que mostró mejores resultados en la minimización del error respecto a otras funciones y la entropía cruzada categórica,  la cual mide la media de bits necesarios para identificar un evento de un conjunto de posibilidades.

Pruebas
Se realizaron un total de 500 validaciones (500 imágenes) en las cuales se presentaron resultados de precisión arriba del 90%, en específico con el 93.8% de probabilidad de acierto. Se realizaron en total 14 entrenamientos variando el número de épocas, el tamaño del lote de imágenes así como los tipos de funciones de activación tales como ‘Rectified Linear Unit (ReLU)  diferentes tipos de funciones de  optimización, tales como Stochastic Gradient Descent (SGD) y el RMSProp. En todos los casos el grado de precisión fue menor, por ejemplo al utilizar RMSProp se obtuvo un 93.2% de acierto, con ReLU un 90% y  SGD con un 88.2 %. De igual manera,  al variar el número de lotes para iterar, se notó que al disminuir el número se disminuía la precisión




history.txt
{'loss': [1.7617307380835214, 1.3895007034142812, 1.250089396238327, 1.1123373119036357, 0.9842512849966685, 0.9192129786809285, 0.8545784181356431, 0.7617239270607631, 0.7146021672089895, 0.668827717701594, 0.6337298866113027, 0.6034104685982068, 0.5829130572080612, 0.5623707944154739, 0.5326512252291044, 0.5169553451736768, 0.49245720585187275, 0.4689043773214022, 0.45053906639417013, 0.5007022746404012, 0.4269529281059901, 0.464196319381396, 0.4244592270751794, 0.39692153612772624, 0.3682126623392105, 0.3719285043080648, 0.38247372855742773, 0.3747043239076932, 0.37083292320370675, 0.4266230917473634, 0.3603544160226981, 0.34316324586669605, 0.3336130006114642, 0.3574215774734815, 0.33076921130220094, 0.350146774550279, 0.3283090650041898, 0.3041694092253844, 0.3386432884633541, 0.3885184265176455, 0.2866205582519372, 0.28318034410476683, 0.2923268772164981, 0.3426624370117982, 0.3256735741098722, 0.3003328355153402, 0.2715418450037638, 0.28015890111525854, 0.29306274066368737, 0.2728440312544505, 0.2581948680679003, 0.26891209507981934, 0.2908134411275387, 0.2762138315786918, 0.271688525006175, 0.2726593572894732, 0.2835807827115059, 0.26059920743107795, 0.26308514674504596, 0.2664711051682631, 0.2618372681240241, 0.2575043226530154, 0.2676563739279906, 0.2897445873916149, 0.27251969491442046, 0.29101292674740153, 0.2586095897356669, 0.23873478417595229, 0.2692802386234204, 0.23818109035491944, 0.24984324951966605, 0.2574720734357834, 0.22555647696057957, 0.24250430022676786, 0.2536205149938663, 0.24495420748988786, 0.29453359723091127, 0.23366982457538446, 0.26688824149469537, 0.2565566722303629, 0.2689083250860373, 0.22452449979881445, 0.21601748724778494, 0.23724826777974764, 0.24501697537799677, 0.23802721490462622, 0.22119592169920604, 0.23494780180354913, 0.2139000395933787, 0.21837289728224277, 0.2418299771596988, 0.22617073446512223, 0.21181476327280205, 0.21306127967933813, 0.2037917010486126, 0.20321005200346312, 0.23402322928110758, 0.2754972216238578, 0.21809080436825753, 0.2195727208877603, 0.20765654580046733, 0.2073758082340161, 0.22751080627242723, 0.21591652803122996, 0.22407936538259188, 0.1994855552415053, 0.19076141498982907, 0.20160903905828795, 0.19232961046199004, 0.19753092187146346, 0.1996143456051747, 0.22414905341962973, 0.24268103561053672, 0.21083850858112177, 0.23575036582847436, 0.2530791735649109, 0.19026481007536253, 0.22825584980348745, 0.26243556479612984, 0.18390756037086248, 0.22018258870889743, 0.24175542819003265, 0.1929008647426963, 0.1853866515060266, 0.20589903031786283, 0.19337248735129833, 0.1795508173108101, 0.19679575095574062, 0.19902543338636558, 0.18117795921862126, 0.1926404257118702, 0.1807290062556664, 0.2222325104971727, 0.1833548025538524, 0.18723474038143953, 0.1853943072259426, 0.1801232883085807, 0.19302764013409615, 0.1771324878682693, 0.2158975794663032, 0.20400121971964835, 0.17501473399500053, 0.16353998565425476, 0.1753036932150523, 0.17492368654658397, 0.21472870747248332, 0.22144277262190978, 0.17460535578429698, 0.19022126264870168, 0.2310796818882227, 0.18809473746766647, 0.17169411968439818, 0.1694001271451513, 0.18462007152537505, 0.18079804030557473, 0.18222737884769838, 0.17339760075012842, 0.18798306196928025, 0.1965165213868022, 0.1682129195332527, 0.1643124252309402, 0.16267868982007105, 0.2185232095917066, 0.1739164814228813, 0.16658080826203028, 0.18311540178954602, 0.16746471072236696, 0.15129163143535454, 0.16749340986212094, 0.17975924858202538, 0.1598579076708605, 0.15557553932070733, 0.17642073767880598, 0.1942849033822616, 0.23668526971091827, 0.1607716754078865, 0.15154859922826291, 0.16229443752517303, 0.16419398953517278, 0.18429300477107366, 0.16586671061813832, 0.1578098430360357, 0.16912953836222488, 0.19616030090798933, 0.23190626693268618, 0.18787157913049063, 0.15183781829973061, 0.1603526403506597, 0.1833548194418351, 0.1808326743915677, 0.17314558568100136, 0.13596066688497863, 0.15941428679972888, 0.14164497156937916, 0.13849557953576247, 0.13994639101127784, 0.1396536501745383, 0.16580825525025528, 0.142414859123528, 0.1508120638380448, 0.1356903874501586, 0.1610050155594945, 0.19890664900342622, 0.18205685404439767, 0.18595065594961246, 0.20742038048803807, 0.19490473061800004, 0.15450203228121004, 0.14361834079027175, 0.13878394213815531, 0.13862999402607482, 0.15208812452852727, 0.1291631835947434, 0.13876713265975316, 0.13649654823044935, 0.14990708818038304, 0.16559982457508643, 0.16475609317421913, 0.1427447679763039, 0.13224923998117447, 0.13712808006753524, 0.16654176799580456, 0.13633966175218423, 0.16185794569551945, 0.19306056473404168, 0.17010979091593376, 0.15775348501900832, 0.16782562812169394, 0.1575041735296448, 0.16240193645159404, 0.13898670463512341, 0.12882842503488065, 0.13442794072131317, 0.13824918175737064, 0.12433688901364803, 0.12426522461076578, 0.13965132466206948, 0.1637560216213266, 0.16852224757273992, 0.151088725614051, 0.13526543106883765, 0.20351549309988817, 0.14909839659929275, 0.14512881454080342, 0.12442192580550909, 0.13220183679834008, 0.1381769413004319, 0.15852602587391934, 0.1548299768815438, 0.1293643603598078, 0.1334902805959185, 0.12630044241746266, 0.11656407677878936, 0.13310106195509433, 0.1915055251866579, 0.1389636342599988, 0.11118574474006891, 0.13906671069562435, 0.11732958319286506, 0.11655759993940591, 0.13753430345406134, 0.14749727693696818, 0.1115067133679986, 0.13104642765596508, 0.12366899266839028, 0.12415329795951645, 0.14310436011602481, 0.12386781415591637, 0.1452482842716078, 0.11606054501918454, 0.11700061206395428, 0.13187859143440922, 0.12534557413309813, 0.11112848943720262, 0.10467593712111314, 0.1013600238164266, 0.103156074937433, 0.1058809267791609, 0.11419749593983095, 0.11397275070349376, 0.10020850756516059, 0.1226489907875657, 0.10772852573543787, 0.13661138678590456, 0.12993199354658524, 0.13500936411321163, 0.13305102470020452, 0.12566378115986784, 0.11394481292615334, 0.10393616897364458, 0.11744329502029965, 0

model_04.py
##.-*-coding: utf-8-*---------------------------------------------------------------------------
# Facultad de Matemáticas de la Universidad Autónoma de Yucatán
#--------------------------------------------------------------------------
#
# MODULE: Building a model
# FILE: model_3.py
# USE: Build de Neural Network Arquitecture
#
#> @author
#> Astrid Giselle Rodriguez Us
#
#  DESCRIPTION:
# This script implements a neural network arquitecture using Tensorflow and Keras
# 29 April 2019 - Initial Version
# -- April 2019 - Final Version
# TODO_dd_mmm_yyyy - TODO_describe_appropriate_changes - TODO_name
#
# Modifications:
#22 April 2019 - implemented getLabel function
#01 May   2019 - added graphics to visualize the data
#05 May   2019- Training test and losses extraction
#--------------------------------------------------------------------------
from loadShuffleData_03 import*
import numpy as np
import tensorflow as tf
from tensorflow.keras import layers
from tensorflow import keras
import  matplotlib.pyplot as plt
from random import shuffle
import random
import csv
from keras.models import load_model
import json


def runModel():                                       #Building a simple model

    '''
    model = keras.Sequential([
        keras.layers.Flatten(input_shape=(32, 32)),
        keras.layers.Dense(128, activation=tf.nn.sigmoid),
        keras.layers.Dense(5, activation=tf.nn.softmax)
    ])
    '''

    model = keras.Sequential([
        keras.layers.Flatten(input_shape=(32, 32)),
        keras.layers.Dense(128, activation='relu'),
        keras.layers.Dense(5, activation=tf.nn.softmax)
    ])


    '''
    model = keras.Sequential([
        keras.layers.Flatten(input_shape=(32, 32)),
        keras.layers.Dense(128, activation=tf.nn.sigmoid),
        keras.layers.Dense(5, activation=tf.nn.softmax)
    ])
    '''


    '''
    model = keras.Sequential([
        keras.layers.Flatten(input_shape=(32, 32)),
        keras.layers.Dense(1024, activation=tf.nn.sigmoid),
        keras.layers.Dense(1024, activation=tf.nn.sigmoid),
        keras.layers.Dense(5, activation=tf.nn.softmax)
    ])
    '''
    '''
    model = keras.Sequential([
        keras.layers.Flatten(input_shape=(32, 32)),
        keras.layers.Dense(128, activation= 'sigmoid'),
        keras.layers.Dense(5, activation=tf.nn.softmax)
    ])
    '''


    return model

#------------Train the model---------------------#
def training(modelFile):
    '''
    model.compile(optimizer= 'Adam',
                  loss= tf.losses.sigmoid_cross_entropy,
                  metrics=['accuracy'])
    '''

    model.compile(optimizer= 'Adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])



    '''
    model.compile(tf.train.AdamOptimizer(0.01),
                  loss='sparse_categorical_crossentropy',
                  metrics=['accuracy'])
    '''

    #model.fit(train_images, train_labels, batch_size=50, epochs=500)
    #model.fit(train_images, train_labels, batch_size=10, epochs=1000)
    #model.fit(train_images, train_labels, batch_size=1, epochs=1000)
    history= model.fit(train_images, train_labels, batch_size=55, epochs=1000)

    # ---------------------------Saving the model-------------------------------------#
    model.save(modelFile)
    #model.save_history('losses', 'True','h5')
    #model.to_json(**kwargs)
    print(history.history)
    return history

def loadModel(modelFile):
    # -------------------Load model------------------------#
    model = keras.models.load_model(modelFile)

    model.save_weights('weights',True,'h5')
    return  model

def predict(model, images,labels ):

    # ------------------Evaluate accuracy----------------------#
    test_loss, test_acc = model.evaluate(images, labels)

    print('Test accuracy:', test_acc)
    predictions = model.predict(images)

    return predictions

def dispPredict(model, predictions, num_img):       #Display Prediction
    map_pred= []

    #predictions= predictions*10
    label_0 = val_labels[num_img]
    classes_0 = val_classes[num_img]

    #print(len(predictions[0]))

    predictions[num_img]= np.array(predictions[num_img])
    print(predictions[num_img])
    #print(np.argmin(predictions[0]))

    print('label: ', label_0)
    print('class: ', classes_0)
    print('prediction: ', np.argmax(predictions[num_img]))

    plt.figure(figsize=(6, 3))
    plt.subplot(1, 2, 1)
    predict = plot_image(num_img, predictions, val_labels, val_images)
    plt.subplot(1, 2, 2)
    plot_value_array(num_img, predictions, val_labels)
    plt.show()

    map_pred.append(num_img)
    map_pred.append(predict)
    return map_pred


def saveWrongPred(predictions, labels):
    result=[]
    flag= 1
    contFalse= 0                                    #Counts the number of wrong predictions
    wrongPImg= []                                   #Saves the index of the wrong predictions
    for i in range(0,500):
        #print('n: ', i)
        if np.argmax(predictions[i]) != labels[i]:
            flag=0
            contFalse= contFalse+1
            wrongPImg.append(i)
        result.append(flag)
        flag=1
        
    print('Number of wrong predictions: ', contFalse)
    # print('Images of wrong predictions: ', wrongPImg)
    return wrongPImg

def a_summary(predictions, labels):                #Returns the correct and wrong predictions of a's
    result = []
    flag = 1
    contFalse = 0  # Counts the number of wrong predictions
    wrongPImg = []  # Saves the index of the wrong predictions
    correctPImg =  []  # Saves the index of the correct predictions
    for i in range(len(predictions)):
        # print('n: ', i)
        if labels[i]==0:
            if np.argmax(predictions[i]) != labels[i]:
                flag = 0
                contFalse = contFalse + 1
                wrongPImg.append(i)

            else:
                correctPImg.append(i)
            result.append(flag)
            flag = 1


    print('Number of wrong predictions: ', contFalse)
    # print('Images of wrong predictions: ', wrongPImg)
    return correctPImg,wrongPImg                #             #

def e_summary(predictions, labels):                #Returns the correct and wrong predictions of e's
    result = []
    flag = 1
    contFalse = 0  # Counts the number of wrong predictions
    wrongPImg = []  # Saves the index of the wrong predictions
    correctPImg =  []  # Saves the index of the correct predictions
    for i in range(len(predictions)):
        # print('n: ', i)
        if labels[i]==1:
            if np.argmax(predictions[i]) != labels[i]:
                flag = 0
                contFalse = contFalse + 1
                wrongPImg.append(i)

            else:
                correctPImg.append(i)
            result.append(flag)
            flag = 1


    print('Number of wrong predictions: ', contFalse)
    # print('Images of wrong predictions: ', wrongPImg)
    return correctPImg,wrongPImg                #             #

def i_summary(predictions, labels):                #Returns the correct and wrong predictions of i's
    result = []
    flag = 1
    contFalse = 0  # Counts the number of wrong predictions
    wrongPImg = []  # Saves the index of the wrong predictions
    correctPImg =  []  # Saves the index of the correct predictions
    for i in range(len(predictions)):
        # print('n: ', i)
        if labels[i]==2:
            if np.argmax(predictions[i]) != labels[i]:
                flag = 0
                contFalse = contFalse + 1
                wrongPImg.append(i)

            else:
                correctPImg.append(i)
            result.append(flag)
            flag = 1


    print('Number of wrong predictions: ', contFalse)
    # print('Images of wrong predictions: ', wrongPImg)
    return correctPImg,wrongPImg                #             #

def o_summary(predictions, labels):                #Returns the correct and wrong predictions of o's
    result = []
    flag = 1
    contFalse = 0  # Counts the number of wrong predictions
    wrongPImg = []  # Saves the index of the wrong predictions
    correctPImg =  []  # Saves the index of the correct predictions
    for i in range(len(predictions)):
        # print('n: ', i)
        if labels[i]==3:
            if np.argmax(predictions[i]) != labels[i]:
                flag = 0
                contFalse = contFalse + 1
                wrongPImg.append(i)

            else:
                correctPImg.append(i)
            result.append(flag)
            flag = 1


    print('Number of wrong predictions: ', contFalse)
    # print('Images of wrong predictions: ', wrongPImg)
    return correctPImg,wrongPImg                #             #

def u_summary(predictions, labels):                #Returns the correct and wrong predictions of u's
    result = []
    flag = 1
    contFalse = 0  # Counts the number of wrong predictions
    wrongPImg = []  # Saves the index of the wrong predictions
    correctPImg =  []  # Saves the index of the correct predictions
    for i in range(len(predictions)):
        # print('n: ', i)
        if labels[i]==4:
            if np.argmax(predictions[i]) != labels[i]:
                flag = 0
                contFalse = contFalse + 1
                wrongPImg.append(i)

            else:
                correctPImg.append(i)
            result.append(flag)
            flag = 1


    print('Number of wrong predictions: ', contFalse)
    # print('Images of wrong predictions: ', wrongPImg)
    return correctPImg,wrongPImg                #             #

#----------------------Train dataset--------------------------------------------#

fileCSV_train= 'shuffle_train.cvs'
fileNPY_train= 'shuffleIMG_train.npy'
numImg_train = 1500

#------------ shuffle-------------------------------------------------------------#
#trainData= load_TrainData(numImg_train)
#wShuffle_train= w_shuffle_data(trainData, fileNPY_train,fileCSV_train)

dataset_train= r_shuffle_data(fileNPY_train,fileCSV_train,numImg_train)

#------------Plotting dataset train----------------------------------------------#
#plot_dataset(dataset_train, numImg_train)

train_images= [row[0] for row in dataset_train]
train_labels= [row[1] for row in dataset_train]
train_classes= [row[2] for row in dataset_train]


train_images= np.array(train_images)
print(train_images.shape)
#print(train_classes[0])

#---------------------Validation dataset------------------------------------------#

fileCSV_val= 'shuffle_val.cvs'
fileNPY_val= 'shuffleIMG_val.npy'
numImg_val= 500
#------------ shuffle-------------------------------------------------------------#
#valData= load_valData(numImg_val)
#wShuffle_val = w_shuffle_data(valData, fileNPY_val, fileCSV_val)

dataset_val= r_shuffle_data(fileNPY_val,fileCSV_val,numImg_val)

#------------Plotting dataset val----------------------------------------------#
#plot_dataset(dataset_val, numImg_val)

val_images= [row[0] for row in dataset_val]
val_labels= [row[1] for row in dataset_val]
val_classes= [row[2] for row in dataset_val]

val_images= np.array(val_images)
print(val_images.shape)

#---------------------Test dataset------------------------------------------#


fileCSV_test= 'shuffle_test.cvs'
fileNPY_test= 'shuffleIMG_test.npy'
'''
#------------ shuffle-------------------------------------------------------------#
#valData= load_valData(numImg_val)
#wShuffle_val = w_shuffle_data(valData, fileNPY_val, fileCSV_val)
'''
numImg_test= 500


#dataset_test= load_testData(numImg_test)
dataset_test= r_shuffle_data(fileNPY_test,fileCSV_test,numImg_val)

#------------Plotting dataset val----------------------------------------------#
#plot_dataset(dataset_val, numImg_val)

test_images= [row[0] for row in dataset_test]
test_labels= [row[1] for row in dataset_test]
test_classes= [row[2] for row in dataset_test]

test_images= np.array(test_images)
print('test_data', test_images.shape)
#test_labels = [int(i) for i in test_labels]
print(len(test_labels))



#------------------------------------------------------------------------------------#


modelFile= 'model_arch.h5'
#modelFile= 'model_arch_2.h5'
#modelFile= 'model_arch_3.h5'
#modelFile= 'model_arch_4.h5'
#modelFile= 'model_arch_5.h5'
#modelFile= 'model_arch_6.h5'
#modelFile= 'model_arch_7.h5'
#modelFile= 'model_arch_8.h5'
#modelFile = 'model_arch_9.h5'
#modelFile = 'model_arch_10.h5'
#modelFile = 'model_arch_11.h5'
#modelFile = 'model_arch_12.h5'                                  #Adam Optimizer -  0.938 accuracy
#modelFile = 'model_arch_13.h5'                                  #SGD Optimizer - 0.882 accuracy
#modelFile = 'model_arch_14.h5'                                   #RMSProp Optimzer - 0.932 accuracy
#modelFile = 'model_arch_15.h5'                                   #Relu activation   - 0.90 accuracy

'''
model= runModel()
history = training(modelFile)
                                       
f = open("history.txt","w")                                     #Save the losses an accuracy data 
f.write( str(history.history) )
f.close()
'''



model = loadModel(modelFile)

#predictions= predict(model, train_images,train_labels)         #predictions for train images
#predictions= predict(model, val_images,val_labels)              #predictions for validation images
predictions= predict(model, test_images,test_labels)            #predictions for test images
wrongPImg= saveWrongPred(predictions,test_labels)
#print(wrongPImg)
aSumC, aSumW= a_summary(predictions,test_labels)

print('Correct A: ', len(aSumC))
print('Incorrect A: ', len(aSumW))

eSumC, eSumW = e_summary(predictions,test_labels)

print('Correct E: ', len(eSumC))
print('Incorrect E: ', len(eSumW))

iSumC, iSumW = i_summary(predictions,test_labels)

print('Correct I: ', len(iSumC))
print('Incorrect I: ', len(iSumW))

oSumC, oSumW = o_summary(predictions,test_labels)

print('Correct 0: ', len(oSumC))
print('Incorrect 0: ', len(oSumW))

uSumC, uSumW = u_summary(predictions,test_labels)

print('Correct U: ', len(uSumC))
print('Incorrect U: ', len(uSumW))


#---------------------------Displaying A predictions
for item in uSumW:
    print(item)
    dispPredict(model,predictions, item)






'''
#---------------------------Displaying wrong predictions
for item in wrongPImg:
    dispPredict(model,predictions, item)

'''



